### PR TITLE
[SPARK-21563][CORE] Fix race condition when serializing TaskDescriptions and adding jars

### PR DESCRIPTION
**Pull request by ash211**
## What changes were proposed in this pull request?

Fix the race condition when serializing TaskDescriptions and adding jars by keeping the set of jars and files for a TaskSet constant across the lifetime of the TaskSet.  Otherwise TaskDescription serialization can produce an invalid serialization when new file/jars are added concurrently as the TaskDescription is serialized.

## How was this patch tested?

Additional unit test ensures jars/files contained in the TaskDescription remain constant throughout the lifetime of the TaskSet.

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/SparkContext.scala
Comment (jiangxb1987) [inline core/src/main/scala/org/apache/spark/SparkContext.scala]: Should we document the behavior?

**Review by jiangxb1987 — APPROVED** [review]
LGTM, cc @cloud-fan

File: core/src/main/scala/org/apache/spark/SparkContext.scala
Comment (ash211) [inline core/src/main/scala/org/apache/spark/SparkContext.scala]: This additional sentence is probably overkill, since I'm sure there are many other subtleties about these methods that aren't spelled out in the Javadoc.

I'm happy to remove and keep these docs simple if that's what a committer would prefer before merging.

**Comment by cloud-fan** [comment]
LGTM, merging to master/2.2!

**Pull request closed without merging**
